ddtrace/internal/_encoding.pyx

```diff
@@ -338,30 +341,96 @@ cdef class Packer(object):
         return buff_to_buff(self.pk.buf, self.pk.length)
 
 
-cdef class MsgpackEncoder(object):
+cdef class BufferedEncoder(object):
+    content_type: str = None
+    def __init__(self, max_size, max_item_size): ...
+    def __len__(self): ...
+    def put(self, trace): ...
+    def encode(self): ...
+
+
+cdef class MsgpackEncoder(BufferedEncoder):
     content_type = "application/msgpack"
+    cdef int _max_size
+    cdef int _max_item_size
+    cdef int _size
+    cdef object _lock
+    cdef list _traces
+
+    def __cinit__(self, max_size, max_item_size):
+        self._max_size = max_size
+        self._max_item_size = max_item_size
+        self._size = 0
+        self._lock = threading.Lock()
+        self._traces = []
+
+    def __len__(self):
+        return len(self._traces)
 
     cpdef _decode(self, data):
         import msgpack
         if msgpack.version[:2] < (0, 6):
             return msgpack.unpackb(data)
         return msgpack.unpackb(data, raw=True)
 
+    @property
+    def max_size(self):
+        return self._max_size
+
+    @property
+    def max_item_size(self):
+        return self._max_item_size
+
+    @property
+    def size(self):
+        """Return the size in bytes of the encoder buffer."""
+        with self._lock:
+            return self._size
+
+    cpdef put(self, list trace):
+        """Put a trace (i.e. a list of spans) in the buffer."""
+        cdef int item_len
+        
+        encoded_trace = Packer().pack_trace(trace)
+        item_len = len(encoded_trace)
+
+        if item_len > self.max_item_size or item_len > self.max_size:
+            raise BufferItemTooLarge(item_len)
+
+        with self._lock:
+            if self._size + item_len <= self.max_size:
+                self._traces.append(encoded_trace)
+                self._size += item_len
+            else:
+                raise BufferFull(item_len)
+
+    cpdef _clear(self):
+        self._traces[:] = []
```

Comment:
Just curious - any difference between this and plain old `self._traces = []`? Just one less object to gc potentially?

Reply:
Yes this should just re-use the same list object